### PR TITLE
Enable caches in deploy task

### PR DESCRIPTION
**Pull request by tdgroot**
Sometimes a cache type gets disabled, for whatever [reason or bug](https://github.com/magento/magento2/issues/17634). 

It's nice to make sure the caches you **_want_** to be enabled **_are_** actually enabled.

This new task makes sure those caches are enabled. The variable `cache_enabled_caches` can be changed to add caches from own or 3rd party modules.

### CODE REVIEW

**Comment by sprankhub** [comment]
I currently added a simple cache enable task, which just enables any cache. Are there use cases where you only want to enable some caches? Shouldn't you want to enable all caches on live?

**Comment by tdgroot** [comment]
@sprankhub thank you for your input. 

I think you're right. Ideally, all caches should be enabled, but I think it should also be possible to enable only certain caches. I've had situations where some caches were meant to be disabled because they were causing more problems than performance gains. 

I can change it so one can set `cache_enabled_caches` to either `all` or an array of specific caches. When `all` is given, all caches are enabled by simply providing no arguments to the `bin/magento cache:enable` command. When specific caches are provided, it will only enable those caches. 

What do you think of this proposal? And what should be the default?

**Comment by sprankhub** [comment]
I think the default should be to enable all caches. Then, an option to define caches, which should not automatically be enabled, which is empty by default, makes the most sense. But this is just my humble opinion :-)

Thanks for your contribution!

**Comment by tdgroot** [comment]
@sprankhub I have updated the pull request. The default is now to enable all the caches. To override that, one can do the following in `deploy.php`:

``` php
<?php

...

// Default setting, enables all caches
set('cache_enabled_caches', 'all');

// Disable enabling of caches
set('cache_enabled_caches', []);

// Only enable specific caches
set('cache_enabled_caches', 
    [
        'config',
        'layout',
        'block_html',
        'collections',
        'reflection',
        'db_ddl',
        'eav',
        'customer_notification',
        'full_page',
        'config_integration',
        'config_integration_api',
        'translate',
        'config_webservice',
        'compiled_config',
    ]
);

...
```

**Comment by osrecio** [comment]
The bug was solved by my teamate https://github.com/magento/magento2/pull/24892. I don't think so that you should a task to enable caches in the main process. 

The tool should be work with all environments and in environments like DEV/STAGING you don't want have all caches enabled. That is something that should manage by environment and the responsability don't be of the deployment.

**Comment by tdgroot** [comment]
> The bug was solved by my teamate [magento/magento2#24892](https://github.com/magento/magento2/pull/24892). I don't think so that you should have a task to enable caches in the main process.

We are still experiencing issues where FPC (read: only FPC) gets disabled after deployments, even on Magento 2.3.4. 

> The tool should be work with all environments and in environments like DEV/STAGING maybe you don't want have all caches enabled. That is something that should manage by environment and the responsability don't be of the deployment.

It is possible to configure certain hosts to not enable all caches, like so:

```
<?php

...

host('staging
    ->hostname('example.com')
    ->set('cache_enabled_caches', []);

...

```

File: recipe/magento_2_1/cache.php [hunk 1]
```diff
@@ -18,3 +18,46 @@
         invoke('cache:clear:magento') :
         writeln('Skipped -> maintenance is not set');
 });
+
+set('cache_enabled_caches', 'all');
```
Comment (osrecio) [inline recipe/magento_2_1/cache.php:22]: Hi @tdgroot I like your PR, 
```suggestion
set('cache_enabled_caches', '');
```

I should not enable all caches by default, for some reasons, maybe some projects don't should have all caches enabled.

In your `deploy.php` you should decide what kind of caches you want enable.

Comment (sprankhub) [reply recipe/magento_2_1/cache.php:22]: IMHO, all caches should be enabled by default. If you do not want to enable some caches, you can still customise this in your `deploy.php`.

Comment (osrecio) [reply recipe/magento_2_1/cache.php:22]: In this case you don’ have the possibility to disable caches in your deploy.php.

And I think that deployer not should be responsible to enable caches of application by default.

We can add the function and add to deploy.php to all, I think works for me.

We have several projects that some caches are not enable, for example some amasty caches.

Greetings!

Comment (tdgroot) [reply recipe/magento_2_1/cache.php:22]: @osrecio @sprankhub Thank you for getting back to this changeset.

I propose the following:

1. In the recipe, caches are disabled by default.
2. In the `deploy.php` template, all caches are configured to be enabled.

Comment (osrecio) [reply recipe/magento_2_1/cache.php:22]: Cool, if you do this approach I will merge ASAP

**Comment by tdgroot** [comment]
@osrecio when will you merge the pull request? I bet there are plenty of devs waiting for this because of the weird side effects.

**Comment by osrecio** [comment]
I hope that will be merged tomorrow

**Comment by osrecio** [comment]
Your PR  is merged @tdgroot 

Release 2.7.0

**Pull request merged**